[Vivante Filter,V10] Added a sub-plugin filter for VIM3/Vivante NPU

leemgs opened the pull request:
This commit is to append the Vivante sub-plugin based on the tensor filter
of NNStreamer to support the Vivante NPU driver.

**Changelog**

* Version 10:
   * Added license clause
   * Improved statements of the nnsreamer.spec file

* Version 1-9:
   * Fixed memory leak (e.g., g_free)
   * FIxed security (e.g., nnstreamer.nanifest) for Tizen packaging
   * doc: Added how to build and run the Vivante filter on Ubuntu 18.04 (ARM64)
   * Updated the invoke function to support various models
     (e.g., Inception, Yolo, Mobile SSD)
   * Added Tizen package script for Tizen platform
   * Fixed incorrect var type (e.g., unsigned int for num_tensors)
   * Added annotations for code maintenance
   * Added run.sh for Ubuntu18.04 aarch64 + VIM3 board
   * Added convert function for vivante tensor data type
   * Added vsi_nn_GetTensor of OpenVX apis to obtain tensor data (input, output)
   * Added info[i].name with dummy data (e.g., tensor id)
   * Removed deprecated variables and functions
   * Updated the vivante filter with "multi-files" structure for re-usability
     - "model=/opt/vivante/model/inception_v3.nb,
       /opt/vivante/model/libinception_v3.so
   * Updated the existing vivante tensor filter based on dlopen() scheme
   * Added a functions to check the dlopen operation to avoid undefined symbol issue
   * Updated annotaiton (e.g., image size is 299:299)
   * Added Vivante functions (common APIs) in the vivante_open function
   * Added Vivante functions (common APIs) in the vivante_invoke function
   * Added Vivante functions (common APIs) in the vivante_close function
   * Update annotation to avoid misunderstanding
   * Fixed typos (GstTensorsInfo)
   * Add todo list (e.g., glib's time functions)
   * Added vsi_nn_pub.h header file to use libraries of acuity-ovxlib-dev tool
   * Added a skeleton of the Vivate tensor filter
   * Imported the major implementation of the Vivante (inceptionv3 model)
     for Nnstreamer pipeline. libvivante_v3.so will be used by "model=***"
     interface to support a generic approach for re-usability.
   * Added vivante_inception_v3.h header file for .so file of a NN model
   * Added a data structure (e.g., vivante_data) to manage Vivante data

Signed-off-by: Geunsik Lim <geunsik.lim@samsung.com>

dongju-chae commented on ext/nnstreamer/tensor_filter/vivante/tensor_filter_subplugin.c:
```diff
@@ -0,0 +1,594 @@
+  pdata->handle = dlopen(pdata->so_path, RTLD_NOW);
+  if (!pdata->handle) {
+    printf ("vivante_open: dlopen cannot load the shared library (.so).\n");
+    return -EINVAL;
```
Possible memory leaks due to line 277~278

leemgs replied:
Do you mean g_free()?

dongju-chae replied:
Yes, you may need `g_free()` for duplicated strings

leemgs replied:
Thanks. Fixed. :)

dongju-chae replied:
Still missing `g_free ()` in line 280.. T_T

myungjoo replied:
@leemgs PTAL here.

leemgs replied:
Ooops. Fixed. :)

leemgs commented:
This PR is ready to be merged.

myungjoo commented on packaging/nnstreamer.spec:
```diff
@@ -36,6 +36,10 @@
 
+%ifarch %arm aarch64
+%define vivante_support 0
+%endif
```
This is weird. Are you leaving it undefined intentionally for other archs?

leemgs replied:
Currently, Vivante supports only ARM32 and ARM64. 
(In case of other archs, when we verify the operation and compatibility, other achs can be added.)

myungjoo commented on packaging/nnstreamer.spec:
```diff
@@ -117,6 +121,18 @@ BuildRequires: lcov
+BuildRequires:  pkgconfig(dann)
+Requires:       gst-plugins-good
+Requires:       gst-plugins-good-extra
+Requires:       gst-libav
```
LINE 131-133 appear incorrect.

leemgs replied:
The three packages are required for the Vivante-based example applications (e.g., custom reference boards) by default.

myungjoo replied:
Goto Line 228

leemgs replied:
Line 131-133 is moved to Line 228.

myungjoo commented on packaging/nnstreamer.spec:
```diff
@@ -117,6 +121,18 @@ BuildRequires: lcov
+BuildRequires:  pkgconfig(opencv)
+BuildRequires:  pkgconfig(ovxlib)
+BuildRequires:  pkgconfig(amlogic-vsi-npu-sdk)
+BuildRequires:  pkgconfig(dann)
```
Do you need dann as well?

leemgs replied:
Let's keep the "**pkgconfig(dann)**"  statement for them temporarily.

leemgs replied:
Instead, I added a "TODO" statement.

myungjoo commented on packaging/nnstreamer.spec:
```diff
@@ -204,6 +220,19 @@ Requires:	armnn
 
+# Support vivante subplugin
+%if 0%{?vivante_support}
+%package -n nnstreamer-vivante
```
drop ```-n nnstreamer-```.

leemgs replied:
Oh, Updated. :)

myungjoo commented on packaging/nnstreamer.spec:
```diff
@@ -204,6 +220,19 @@ Requires:	armnn
+%if 0%{?vivante_support}
+%package -n nnstreamer-vivante
+Summary:    NNStreamer subplugin for Verisilion's Vivante
+Requires:   nnstreamer
```
I'd recommend to add ``` = %{version}-%{release}```.

leemgs replied:
Added.

myungjoo commented on packaging/nnstreamer.spec:
```diff
@@ -204,6 +220,19 @@ Requires:	armnn
+%package -n nnstreamer-vivante
+Summary:    NNStreamer subplugin for Verisilion's Vivante
+Requires:   nnstreamer
+%description -n nnstreamer-vivante
```
Drop ```-n nnstreamer-```

leemgs replied:
Updated.

myungjoo commented on packaging/nnstreamer.spec:
```diff
@@ -608,6 +637,13 @@ cp -r result %{buildroot}%{_datadir}/nnstreamer/unittest/
 %endif  # mvncsdk2_support
 
+%if 0%{?vivante_support}
+%files -n nnstreamer-vivante
```
Drop -n nnstreamer-

leemgs replied:
Updated.

leemgs commented on ext/nnstreamer/tensor_filter/vivante/tensor_filter_subplugin.c:
```diff
@@ -0,0 +1,602 @@
+ * This is a template with no license requirements.
+ * Writers may alter the license to anything they want.
+ * The author hereby allows to do so.
+ */
```
Nice catch. Updated.

myungjoo replied:
This is incorrect. SPDX should be at LINE 1.

myungjoo replied:
It's good now.

leemgs commented:
**Changelog**

* Version 10:
   * Added license clause
   * Improved statements of the nnsreamer.spec file

* Version 1-9:
   * Fixed memory leak (e.g., g_free)
   * FIxed security (e.g., nnstreamer.nanifest) for Tizen packaging
   * doc: Added how to build and run the Vivante filter on Ubuntu 18.04 (ARM64)
   * Updated the invoke function to support various models
     (e.g., Inception, Yolo, Mobile SSD)
   * Added Tizen package script for Tizen platform
   * Fixed incorrect var type (e.g., unsigned int for num_tensors)
   * Added annotations for code maintenance
   * Added run.sh for Ubuntu18.04 aarch64 + VIM3 board
   * Added convert function for vivante tensor data type
   * Added vsi_nn_GetTensor of OpenVX apis to obtain tensor data (input, output)
   * Added info[i].name with dummy data (e.g., tensor id)
   * Removed deprecated variables and functions
   * Updated the vivante filter with "multi-files" structure for re-usability
     - "model=/opt/vivante/model/inception_v3.nb,
       /opt/vivante/model/libinception_v3.so
   * Updated the existing vivante tensor filter based on dlopen() scheme
   * Added a functions to check the dlopen operation to avoid undefined symbol issue
   * Updated annotaiton (e.g., image size is 299:299)
   * Added Vivante functions (common APIs) in the vivante_open function
   * Added Vivante functions (common APIs) in the vivante_invoke function
   * Added Vivante functions (common APIs) in the vivante_close function
   * Update annotation to avoid misunderstanding
   * Fixed typos (GstTensorsInfo)
   * Add todo list (e.g., glib's time functions)
   * Added vsi_nn_pub.h header file to use libraries of acuity-ovxlib-dev tool
   * Added a skeleton of the Vivate tensor filter
   * Imported the major implementation of the Vivante (inceptionv3 model)
     for Nnstreamer pipeline. libvivante_v3.so will be used by "model=***"
     interface to support a generic approach for re-usability.
   * Added vivante_inception_v3.h header file for .so file of a NN model
   * Added a data structure (e.g., vivante_data) to manage Vivante data

wooksong commented on ext/nnstreamer/tensor_filter/vivante/README.md:
```diff
@@ -0,0 +1,166 @@
+ * Read multi-files (e.g., vivante_v3.nb and libvivante_v3.so)
+ * Create a neural network
+ * Execute pre-process for the image data
+ * Verity a graph
```
typo
```suggestion
 * Verify a graph
```

leemgs replied:
Nice catch. Fixed.

wooksong commented on ext/nnstreamer/tensor_filter/vivante/README.md:
````diff
@@ -0,0 +1,166 @@
+```
+
+## How to run
+You can get the below result as follows when the required packages are prepared on the target board.
````
IMHO, 'the below result' and 'as follows' could be a duplicate phrase.

leemgs replied:
Fixed. Thanks.

myungjoo commented on ext/nnstreamer/tensor_decoder/tensordecutil.c:
```diff
@@ -116,3 +116,4 @@ _free_labels (imglabel_t * data)
   data->total_labels = 0;
   data->max_word_length = 0;
 }
+
```
??? Remove this file change from the PR.

leemgs replied:
Fixed.

again4you commented on ext/nnstreamer/tensor_filter/vivante/README.md:
```diff
@@ -0,0 +1,166 @@
+You can get the below result when the required packages are prepared on the target board.
+
+
+#### Case study: Runn inceptionv3 model on the Ubuntu18.04 + VIM3 NPU board
```
typo: `Runn` -> `Run`

leemgs replied:
Fixed. Thanks.

again4you commented on ext/nnstreamer/tensor_filter/vivante/README.md:
````diff
@@ -0,0 +1,166 @@
+-rw-rw-r-- 1 khadas khadas 2002 Jan 30 00:43 ./vivante.out.bin
+```
+
+#### Case study: Runn Yolov3 model on the Ubuntu18.04 + VIM3 NPU board
````
ditto.

leemgs replied:
Fixed. :)

zhoonit commented on ext/nnstreamer/tensor_filter/vivante/README.md:
```diff
@@ -0,0 +1,166 @@
+## How to buid
+First of all, you must generate a library of a model (e.g., libvivantev3.so, libyolov3.so) to use NNStreamer tensor filter.
+For more details, please refer to the below repository.
+ * https://www.github.com/nnstreamer/vivante-nnstreamer-filter (TODO) - Press 'models' folder.
```
I think the address is planned to be changed to `https://github.com/nnstreamer/reference-models`

leemgs replied:
Nice catch. Fixed. :)

zhoonit commented on ext/nnstreamer/tensor_filter/vivante/build_run.sh:
```diff
@@ -0,0 +1,108 @@
+
+# @Author: Geunsik Lim <geunsik.lim@samsung.com>
+# @brief: The simple script is to build and run the VIM3/Vivante fiflter 
+# This script is evaluated on the Ubuntu 18.04 (arm64) + VIM3/Vivante board.
```
Just a suggestion :)
```suggestion
# This script has been tested on the Ubuntu 18.04 (arm64) + VIM3/Vivante board.
```

leemgs replied:
Updated.

helloahn commented on ext/nnstreamer/tensor_filter/vivante/tensor_filter_subplugin.c:
```diff
@@ -0,0 +1,606 @@
+   /** Copy an input buffer to an input tensor */
+    status = call(result_vsi_nn_CopyDataToTensor, VSI_FAILURE, pdata->graph,
+        tensor, input[i].data);
+    //vsi_nn_CopyDataToTensor(pdata->graph, tensor, input[i].data );
```
IMO, this comment can be removed.

leemgs replied:
Removed.

helloahn commented on ext/nnstreamer/tensor_filter/vivante/tensor_filter_subplugin.c:
```diff
@@ -0,0 +1,606 @@
+  if (pdata->postProcess)
+    ret = doPostProcess (pdata);
+  if (ret < 0) {
+    g_printerr("PostProcess Failure\n");
```
Q: go forward after failure?

helloahn commented on ext/nnstreamer/tensor_filter/vivante/tensor_filter_subplugin.c:
```diff
@@ -0,0 +1,606 @@
+ * 1.1. Mandatory APIs:
+ * vnn_CreateNeuralNetwork() to create a neural network from .nb file
+ * vsi_nn_VerifyGraph() to verify the generated graph
+ * vsi_nn_RunGraph() to prcess the graph
```
typo: prcess -> process | run maybe?

leemgs replied:
Nice catch. Fixed. 👍
